File: app/src/main/java/me/kashyap/theverge/ApiModule.java
package me.kashyap.theverge;

import android.content.Context;

import com.squareup.picasso.Picasso;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import me.kashyap.theverge.rest.RssService;
import retrofit.Endpoint;
import retrofit.RestAdapter;
import retrofit.converter.Converter;

/**
 * Created on 3/23/2015.
 */
@Module
public class ApiModule {

    @Singleton
    @Provides
    public RestAdapter providesRestAdapter(Converter converter, Endpoint endpoint) {
        return new RestAdapter.Builder().setLogLevel(RestAdapter.LogLevel.BASIC)
                .setEndpoint(endpoint)
                .setConverter(converter)
                .build();
    }

    @Singleton
    @Provides
    public RssService providesRssService(RestAdapter adapter) {
        return adapter.create(RssService.class);
    }

    @Provides
    @Singleton
    public Picasso providesPicasso(Context context) {
        return Picasso.with(context.getApplicationContext());
    }
}

File: app/src/main/kotlin/me/kashyap/theverge/ApiModule.kt
package me.kashyap.theverge

import android.content.Context
import com.squareup.picasso.Picasso
import dagger.Module
import dagger.Provides
import me.kashyap.theverge.rest.RssService
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Converter
import retrofit2.Retrofit
import retrofit2.RxJavaCallAdapterFactory
import javax.inject.Named
import javax.inject.Singleton

/**
 * Created on 3/23/2015.
 */

const val BASE_URL = "BASE_URL"

@Module
class ApiModule {

    @Singleton
    @Provides
    fun providesRestAdapter(@Named(BASE_URL) baseUrl: String, converterFactory: Converter.Factory): Retrofit {
        var httpLoggingInterceptor = HttpLoggingInterceptor()
        httpLoggingInterceptor.setLevel(HttpLoggingInterceptor.Level.BODY)
        var client = OkHttpClient.Builder()
                .addInterceptor(httpLoggingInterceptor)
        .build()

        return Retrofit.Builder()
                .baseUrl(baseUrl)
                .client(client)
                .addCallAdapterFactory(RxJavaCallAdapterFactory.create())
                .addConverterFactory(converterFactory)
                .build()
    }

    @Singleton
    @Provides
    fun providesRssService(adapter: Retrofit): RssService {
        return adapter.create(RssService::class.java)
    }

    @Provides
    @Singleton
    fun providesPicasso(context: Context): Picasso {
        return Picasso.with(context.applicationContext)
    }
}

